### ISSUE TITLE
RuntimeError: CUDA out of memory.

### ISSUE THREAD

**Issue by menickname** [nanoporetech/bonito#41]
Dear @iiSeymour 

I have been continuing training bonito models with some new datasets. To do so I have used subsets of a mix of bacterial species data (no longer a single dataset), inputting comparable amount of training input (500.000.000 basepairs). When running the Bonito train command on our HPC facility, I do get a RuntimeError: CUDA out of memory. Do you have any estimation on memory consumption/requirement for the Bonito training. I have used the --multi-gpu flag already and tried 2 GPUs, however resulting in the same error message. Our HPC facility has 4x NVIDIA Volta V100 GPUs (32GB GPU memory) on each of 10 nodes.

Thank you in advance.
Best regards,
Nick Vereecke

**Comment by iiSeymour** [comment]
Hey Nick, 

That's odd, what batch size are you trying to use and are you using `--amp`? 

Another thing to check is that the GPUs are free - some applications/libraries will allocate all the memory on all the GPUs on a given node, for example TensorFlow:

> By default, TensorFlow maps nearly all of the GPU memory of all GPUs (subject to CUDA_VISIBLE_DEVICES) visible to the process. This is done to more efficiently use the relatively precious GPU memory resources on the devices by reducing memory fragmentation. 

You can use the `nvidia-smi` command to see which GPUs are running jobs and how much memory is allocated on each.

```
$ nvidia-smi
+-----------------------------------------------------------------------------+
| NVIDIA-SMI 440.33.01    Driver Version: 440.33.01    CUDA Version: 10.2     |
|-------------------------------+----------------------+----------------------+
| GPU  Name        Persistence-M| Bus-Id        Disp.A | Volatile Uncorr. ECC |
| Fan  Temp  Perf  Pwr:Usage/Cap|         Memory-Usage | GPU-Util  Compute M. |
|===============================+======================+======================|
|   0  Tesla V100-PCIE...  On   | 00000000:18:00.0 Off |                    0 |
| N/A   27C    P0    24W / 250W |      0MiB / 16160MiB |      0%      Default |
+-------------------------------+----------------------+----------------------+
|   1  Tesla V100-PCIE...  On   | 00000000:3B:00.0 Off |                    0 |
| N/A   28C    P0    24W / 250W |      0MiB / 16160MiB |      0%      Default |
+-------------------------------+----------------------+----------------------+
|   2  Tesla V100-PCIE...  On   | 00000000:86:00.0 Off |                    0 |
| N/A   27C    P0    23W / 250W |      0MiB / 16160MiB |      0%      Default |
+-------------------------------+----------------------+----------------------+
|   3  Tesla V100-PCIE...  On   | 00000000:AF:00.0 Off |                    0 |
| N/A   28C    P0    24W / 250W |      0MiB / 16160MiB |      0%      Default |
+-------------------------------+----------------------+----------------------+
                                                                               
+-----------------------------------------------------------------------------+
| Processes:                                                       GPU Memory |
|  GPU       PID   Type   Process name                             Usage      |
|=============================================================================|
|  No running processes found                                                 |
+-----------------------------------------------------------------------------+
```

**Comment by menickname** [comment]
Dear @iiSeymour 

Thanks for you quick response. I do indeed use the --amp option in my command. Let me check with the HPC responsible for the TensorFlow allocation and occupancy of GPUs.

Thanks

**Comment by boegel** [comment]
The GPUs are "reserved" for the jobs that request them, so it should be impossible that the GPU memory is being used by another process when @menickname is running Bonito on one or more GPUs.

The full error we're seeing looks like this:

```
RuntimeError: CUDA out of memory. Tried to allocate 364.00 MiB (GPU 0; 31.75 GiB total capacity; 29.83 GiB already allocated; 143.69 MiB free; 30.58 GiB reserved in total by PyTorch)
```

So PyTorch (which is used by Bonito, I think?) has already reserved the GPU memory, but then there's insufficient memory for Bonito to allocate? Or how should we interpret this?

**Comment by iiSeymour** [comment]
Yes libraries like PyTorch and TensorFlow will over allocated memory to avoid fragmentation. If the GPUs are reserved then the only two things I can think of are:

 - the batch size is too large.
 - the chunk size is too large when creating the training data.

**Comment by menickname** [comment]
Dear @iiSeymour 

I will try to lower the batch size first and see what this gives.

Thanks in advance.

**Comment by iiSeymour** [comment]
Hey Nick, did you get to the bottom of this?

**Comment by menickname** [comment]
hey @iiSeymour , I have lowered the batch size and I succeeded in getting the job done. However, it took more than 3 days to get the dataset finished, so I will still look further into this how to speed up the process. To my knowledge only 1 GPU was used now for this analysis.

**Comment by iiSeymour** [comment]
That's sound about right for a single GPU for now - I hope to release future models that coverage much faster.

**Comment by menickname** [comment]
Great! looking forward to hear from that! :D Thanks a lot!

**Comment by iiSeymour** [comment]
Hey @menickname,

I've just released [v0.2.2](https://github.com/nanoporetech/bonito/releases/tag/v0.2.2) which has a bunch of training improvements - in my tests on a single GPU using batch size 32, I find a difference of > 1% after training for 20 epochs.

**Comment by boegel** [comment]
@menickname Bonito 0.2.2 now installed on `joltik`
